File: csharp/Betfair.ESAClient/Betfair.ESAClient/ClientCache.cs
using Betfair.ESAClient.Cache;
using Betfair.ESAClient.Protocol;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Betfair.ESASwagger.Model;

namespace Betfair.ESAClient
{
    /// <summary>
    /// Simple ESA Cache implementation that wraps an ESA Client
    /// and caches the streams of data.
    /// </summary>
    public class ClientCache : IChangeMessageHandler
    {
        private readonly MarketCache _marketCache = new MarketCache();
        private readonly OrderCache _orderCache = new OrderCache();
        private readonly Client _client;

        /// <summary>
        /// Construct a new cache to consume from / wrap the specified client.
        /// </summary>
        /// <param name="client"></param>
        public ClientCache(Client client)
        {
            _client = client;
            _client.ChangeHandler = this;
        }

        /// <summary>
        /// The underlying Client
        /// </summary>
        public Client Client
        {
            get
            {
                return _client;
            }
        }

        /// <summary>
        /// The underlying Client status
        /// </summary>
        public ConnectionStatus Status
        {
            get
            {
                return _client.Status;
            }
        }
            
        /// <summary>
        /// Explicitly start the client (otherwise this is automatic on first subscribe).
        /// </summary>
        public void Start()
        {
            _client.Start();
        }

        /// <summary>
        /// Explicitly stop the client
        /// </summary>
        public void Stop()
        {
            _client.Stop();
        }

        /// <summary>
        /// Set a common market data filter for all market subscriptions
        /// </summary>
        public MarketDataFilter MarketDataFilter
        {
            get
            {
                return _client.MarketDataFilter;
            }
            set
            {
                _client.MarketDataFilter = value;
            }
        }

        /// <summary>
        /// Set a common requested conflation rate (this slows rate of changes down) in milliseconds.
        /// </summary>
        public long? ConflatMs
        {
            get
            {
                return _client.ConflateMs;
            }
            set
            {
                _client.ConflateMs = value;
            }
        }

        /// <summary>
        /// Subscribe to all orders. (starting the client if needed).
        /// </summary>
        public void SubscribeOrders()
        {
            SubscribeOrders(new OrderSubscriptionMessage());
        }

        /// <summary>
        /// Explict order subscription. (starting the client if needed).
        /// </summary>
        /// <param name="subscription"></param>
        public void SubscribeOrders(OrderSubscriptionMessage subscription)
        {
            _client.Start();
            _client.OrderSubscription(subscription);
        }

        /// <summary>
        /// Subscribe to the specified market ids. (starting the client if needed).
        /// </summary>
        /// <param name="markets"></param>
        public void SubscribeMarkets(params string[] markets)
        {
            SubscribeMarkets(new MarketFilter { MarketIds = new List<string>(markets) });
        }

        /// <summary>
        /// Subscribe to the specified markets (matching your filter). (starting the client if needed).
        /// </summary>
        /// <param name="marketFilter"></param>
        public void SubscribeMarkets(MarketFilter marketFilter)
        {
            SubscribeMarkets(new MarketSubscriptionMessage()
            {
                MarketFilter = marketFilter,
            });
        }

        /// <summary>
        /// Explicit order subscripion. (starting the client if needed).
        /// </summary>
        /// <param name="subscription"></param>
        public void SubscribeMarkets(MarketSubscriptionMessage subscription)
        {
            _client.Start();
            _client.MarketSubscription(subscription);
        }

        /// <summary>
        /// The cache of all subscribed markets
        /// </summary>
        public MarketCache MarketCache
        {
            get
            {
                return _marketCache;
            }
        }

        /// <summary>
        /// The cache of all subscribed orders
        /// </summary>
        public OrderCache OrderCache
        {
            get
            {
                return _orderCache;
            }
        }


        #region IChangeMessageHandler

        void IChangeMessageHandler.OnMarketChange(ChangeMessage<MarketChange> change)
        {
            _marketCache.OnMarketChange(change);
        }

        void IChangeMessageHandler.OnOrderChange(ChangeMessage<OrderMarketChange> change)
        {
            _orderCache.OnOrderChange(change);
        }

        void IChangeMessageHandler.OnErrorStatusNotification(StatusMessage message)
        {

        }

        #endregion

    }
}

File: java/client/src/main/java/com/betfair/esa/client/ClientCache.java
package com.betfair.esa.client;

import com.betfair.esa.client.auth.InvalidCredentialException;
import com.betfair.esa.client.cache.market.MarketCache;
import com.betfair.esa.client.cache.order.OrderCache;
import com.betfair.esa.client.protocol.ChangeMessage;
import com.betfair.esa.client.protocol.ChangeMessageHandler;
import com.betfair.esa.client.protocol.ConnectionException;
import com.betfair.esa.client.protocol.ConnectionStatus;
import com.betfair.esa.client.protocol.ConnectionStatusListener;
import com.betfair.esa.client.protocol.StatusException;
import com.betfair.esa.swagger.model.MarketChange;
import com.betfair.esa.swagger.model.MarketDataFilter;
import com.betfair.esa.swagger.model.MarketFilter;
import com.betfair.esa.swagger.model.MarketSubscriptionMessage;
import com.betfair.esa.swagger.model.OrderMarketChange;
import com.betfair.esa.swagger.model.OrderSubscriptionMessage;
import com.betfair.esa.swagger.model.StatusMessage;
import java.util.Arrays;

/** Created by mulveyj on 07/07/2016. */
public class ClientCache implements ChangeMessageHandler {
    private final MarketCache marketCache = new MarketCache();
    private final OrderCache orderCache = new OrderCache();
    private final Client client;

    public ClientCache(Client client) {
        this.client = client;
        client.setChangeHandler(this);
    }

    public void addConnectionStatusListener(ConnectionStatusListener listener) {
        client.addConnectionStatusListener(listener);
    }

    public void removeConnectionStatusListener(ConnectionStatusListener listener) {
        client.removeConnectionStatusListener(listener);
    }

    public ConnectionStatus getStatus() {
        return client.getStatus();
    }

    public void setHeartbeatMs(Long heartbeatMs) {
        client.setHeartbeatMs(heartbeatMs);
    }

    public Long getHeartbeatMs() {
        return client.getHeartbeatMs();
    }

    public void setConflateMs(Long conflateMs) {
        client.setConflateMs(conflateMs);
    }

    public Long getConflateMs() {
        return client.getConflateMs();
    }

    public void setMarketDataFilter(MarketDataFilter marketDataFilter) {
        client.setMarketDataFilter(marketDataFilter);
    }

    public MarketDataFilter getMarketDataFilter() {
        return client.getMarketDataFilter();
    }

    public Client getClient() {
        return client;
    }

    public MarketCache getMarketCache() {
        return marketCache;
    }

    public OrderCache getOrderCache() {
        return orderCache;
    }

    /** Subscribe to all orders. (starting the client if needed). */
    public void subscribeOrders()
            throws InvalidCredentialException, StatusException, ConnectionException {
        subscribeOrders(new OrderSubscriptionMessage());
    }

    /** Explict order subscription. (starting the client if needed). */
    public void subscribeOrders(OrderSubscriptionMessage subscription)
            throws InvalidCredentialException, StatusException, ConnectionException {
        client.start();
        client.orderSubscription(subscription);
    }

    /** Subscribe to the specified market ids. (starting the client if needed). */
    public void subscribeMarkets(String... markets)
            throws InvalidCredentialException, StatusException, ConnectionException {
        MarketFilter marketFilter = new MarketFilter();
        marketFilter.setMarketIds(Arrays.asList(markets));
        subscribeMarkets(marketFilter);
    }

    /**
     * Subscribe to the specified markets (matching your filter). (starting the client if needed).
     */
    public void subscribeMarkets(MarketFilter marketFilter)
            throws InvalidCredentialException, StatusException, ConnectionException {
        MarketSubscriptionMessage marketSubscriptionMessage = new MarketSubscriptionMessage();
        marketSubscriptionMessage.setMarketFilter(marketFilter);
        subscribeMarkets(marketSubscriptionMessage);
    }

    /// <summary>
    /// Explicit order subscripion. (starting the client if needed).
    /// </summary>
    /// <param name="subscription"></param>
    public void subscribeMarkets(MarketSubscriptionMessage subscription)
            throws InvalidCredentialException, StatusException, ConnectionException {
        client.start();
        client.marketSubscription(subscription);
    }

    @Override
    public void onOrderChange(ChangeMessage<OrderMarketChange> change) {
        orderCache.onOrderChange(change);
    }

    @Override
    public void onMarketChange(ChangeMessage<MarketChange> change) {
        marketCache.onMarketChange(change);
    }

    @Override
    public void onErrorStatusNotification(StatusMessage message) {}
}
